File: cpp/medium/Valid_Sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<vector<int>> index_list;
        int i = 0, j = 0, num = 0;
        for (j=0; j < 9; ++j){
            for (i=0; i < 9; ++i){
                if (board[j][i] != '.'){
                    for (int n=0; n < num; ++n){
                        if (board[j][i] == index_list[n][0] and (i == index_list[n][1] or j == index_list[n][2] or i/3+j/3*3 == index_list[n][3])) return false;
                    }
                    index_list.push_back({board[j][i], i, j, i/3+j/3*3});
                    ++num;
                }
            }
        }
        return true;
    }
};

File: python/medium/Valid_Sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        col_dict = {'1':[], '2':[], '3':[], '4':[], '5':[], '6':[], '7':[], '8':[], '0':[]}
        square_dict = {'1':[], '2':[], '3':[], '4':[], '5':[], '6':[], '7':[], '8':[], '0':[]}
        for row_index in range(len(board)):
            for col_index in range(len(board[row_index])):
                target = board[row_index][col_index]
                square = int(col_index/3) + int(row_index/3)*3
                if target != '.':
                    if board[row_index].count(target) > 1:
                        return False
                    elif target in col_dict[str(col_index)]:
                        return False
                    elif target in square_dict[str(square)]:
                        return False
                    else:
                        col_dict[str(col_index)].append(target)
                        square_dict[str(square)].append(target)
        return True
